torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -28,17 +29,37 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
 
   if (!grads[0].defined())
     return {};
-  if (variable.grad_fn())
+
+  auto var = variable.lock();
+  // It's possible that the Variable went out of scope and was freed.
+  // We still need to handle the unlikely case of someone holding to its grad.
+  if (!var.defined()) {
+    auto var_grad = variable_grad.lock();
+    // Everything was freed. Nothing to do.
+    if (!var_grad.defined()) return variable_list();
+    // Now here's the hard part. If both the new_grad and var_grad require grad
+    // then we just acumulate the data in place (as we'd do if the Variable was
+    // alive). Otherwise, we'd need to perform the out-of-place reduction, but
+    // since the user only holds a reference to .grad and there's no way to
+    // give him the new Value, we just assume that they know these attributes
+    // are changing when using higher order graphs.
+    if (GradMode::is_enabled() && var_grad.requires_grad() && grads[0].requires_grad()) {
```

Comment:
I don't think the translation of `volatile()` to `!requires_grad()` is valid here. We can see that there is trouble because if we read the comment, we see that the condition here doesn't actually match the logic that would have been executed if the variable was alive.

This code in-place updates:

```
import torch
W = torch.randn(10, requires_grad=True)
W.grad = torch.zeros(10)
old_grad = W.grad
x =torch.randn(10, requires_grad=True)
z = W * x
z.backward(torch.ones(z.size()))
print(old_grad)
```

But after your change it will not (because `W.grad` does not require grad)

Reply:
Yes, it does not look right.  Apparently the current code translates

```
 } else if (var->grad->is_volatile) {
    acc_inplace(var->grad, new_grad);
```

to

```
  } else if (!GradMode::is_enabled()) {
...
```

So the latest version of this PR just uses the same code as below that don't need `variable` to be defined.  I've also added a test for `sparse`.